File: javalin_questions/APQuiz01Test.java
/**
INSTRUCTIONS: 
    The APQuiz01Test class provides JUnit tests for the methods 
    in the APQuiz01 class.



COMPILE & EXECUTE & CLEANUP (Java):

     javac     -d out           -cp junit-1.9.3.jar     APQuiz01*.java
     java -jar junit-1.9.3.jar  --class-path "out"      --scan-classpath
     rm -rf out/

COMPILE & EXECUTE & CLEANUP (Java x Kotlin):

     javac     -d out   -cp junit-1.9.3.jar             APQuiz01*.java
     kotlinc    -d out                                  APQuiz01*.kt
     java -jar junit-1.9.3.jar --class-path "./out:$KOTLIN_HOME/lib/kotlin-stdlib.jar" --scan-classpath
     rm -rf out/


TIP:
     - On Windows, use ';' as the classpath separator
     - On Mac/Linux, use ':' as the classpath separator
     - java   `-cp` (Mac/Linux) flag must sepcify absolute paths.
     - kotlin `-cp` (Windows) flag accepts relative paths.
 */

// File: APQuiz01Test.java
// JDK-17: https://developers.redhat.com/products/openjdk/download?sc_cid=701f2000000RWTnAAO
// JUnit: https://mvnrepository.com/artifact/org.junit.jupiter/junit-jupiter/5.10.0
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.Arrays;
import java.util.List;

public class APQuiz01Test {
    
    /**
     * < Q1 >
     * Tests the scoresIncreasing method.
     */
    @Test
    public void testScoresIncreasing() {
        assertTrue(APQuiz01.scoresIncreasing(new int[]{1, 3, 4}));            // -> true
        assertFalse(APQuiz01.scoresIncreasing(new int[]{1, 3, 2}));           // -> false
        assertTrue(APQuiz01.scoresIncreasing(new int[]{1, 1, 4}));            // -> true
        assertTrue(APQuiz01.scoresIncreasing(new int[]{1, 1, 2, 4, 4, 7}));   // -> true
        assertFalse(APQuiz01.scoresIncreasing(new int[]{1, 1, 2, 4, 3, 7}));  // -> false
        assertTrue(APQuiz01.scoresIncreasing(new int[]{-5, 4, 11}));          // -> true
    }

    /**
     * < Q2 >
     * Tests the scores100 method.
     */
    @Test
    public void testScores100() {
        assertTrue(APQuiz01.scores100(new int[]{1, 100, 100}));             // -> true
        assertFalse(APQuiz01.scores100(new int[]{1, 100, 99, 100}));        // -> false
        assertTrue(APQuiz01.scores100(new int[]{100, 1, 100, 100}));        // -> true
        assertFalse(APQuiz01.scores100(new int[]{100, 1, 100, 1}));         // -> false
        assertFalse(APQuiz01.scores100(new int[]{1, 2, 3, 4, 5}));          // -> false
        assertFalse(APQuiz01.scores100(new int[]{1, 2, 100, 4, 5}));        // -> false
    }

    /**
     * < Q3 >
     * Tests the scoresClump method.
     */
    @Test
    public void testScoresClump() {
        assertTrue(APQuiz01.scoresClump(new int[]{3, 4, 5}));           // -> true
        assertFalse(APQuiz01.scoresClump(new int[]{3, 4, 6}));          // -> false
        assertTrue(APQuiz01.scoresClump(new int[]{1, 3, 5, 5}));         // -> true
        assertTrue(APQuiz01.scoresClump(new int[]{2, 4, 5, 6}));         // -> true
        assertFalse(APQuiz01.scoresClump(new int[]{2, 4, 5, 7}));         // -> false
        assertTrue(APQuiz01.scoresClump(new int[]{2, 4, 4, 7}));         // -> true
        assertFalse(APQuiz01.scoresClump(new int[]{3, 3, 6, 7, 9}));      // -> false
        assertTrue(APQuiz01.scoresClump(new int[]{3, 3, 7, 7, 9}));       // -> true
        assertFalse(APQuiz01.scoresClump(new int[]{4, 5, 8}));            // -> false
    }

    /**
     * < Q4 >
     * Tests the scoresAverage method.
     */
    @Test
    public void testScoresAverage() {
        assertEquals(4, APQuiz01.scoresAverage(new int[]{2, 2, 4, 4}));
        assertEquals(4, APQuiz01.scoresAverage(new int[]{4, 4, 4, 2, 2, 2}));
        assertEquals(4, APQuiz01.scoresAverage(new int[]{3, 4, 5, 1, 2, 3}));
        assertEquals(6, APQuiz01.scoresAverage(new int[]{5, 6}));
        assertEquals(5, APQuiz01.scoresAverage(new int[]{5, 4}));
        assertEquals(5, APQuiz01.scoresAverage(new int[]{5, 4, 5, 6, 2, 1, 2, 3}));
    }

    /**
     * < Q5 >
     * Tests the wordsCount method.
     */
    @Test
    public void testWordsCount() {
        assertEquals(2, APQuiz01.wordsCount(new String[]{"a", "bb", "b", "ccc"}, 1));
        assertEquals(1, APQuiz01.wordsCount(new String[]{"a", "bb", "b", "ccc"}, 3));
        assertEquals(0, APQuiz01.wordsCount(new String[]{"a", "bb", "b", "ccc"}, 4));
        assertEquals(2, APQuiz01.wordsCount(new String[]{"xx", "yyy", "x", "yy", "z"}, 1));
        assertEquals(2, APQuiz01.wordsCount(new String[]{"xx", "yyy", "x", "yy", "z"}, 2));
        assertEquals(1, APQuiz01.wordsCount(new String[]{"xx", "yyy", "x", "yy", "z"}, 3));
    }

    /**
     * < Q6 >
     * Tests the wordsFront method.
     */
    @Test
    public void testWordsFront() {
        assertArrayEquals(new String[]{"a"}, APQuiz01.wordsFront(new String[]{"a", "b", "c", "d"}, 1));
        assertArrayEquals(new String[]{"a", "b"}, APQuiz01.wordsFront(new String[]{"a", "b", "c", "d"}, 2));
        assertArrayEquals(new String[]{"a", "b", "c"}, APQuiz01.wordsFront(new String[]{"a", "b", "c", "d"}, 3));
        assertArrayEquals(new String[]{"a", "b", "c", "d"}, APQuiz01.wordsFront(new String[]{"a", "b", "c", "d"}, 4));
        assertArrayEquals(new String[]{"Hi"}, APQuiz01.wordsFront(new String[]{"Hi", "There"}, 1));
        assertArrayEquals(new String[]{"Hi", "There"}, APQuiz01.wordsFront(new String[]{"Hi", "There"}, 2));
    }

    /**
     * < Q7 >
     * Tests the wordsWithoutList method.
     */
    @Test
    public void testWordsWithoutList() {
        assertArrayEquals(new String[]{"bb", "ccc"}, APQuiz01.wordsWithoutList(new String[]{"a", "bb", "b", "ccc"}, 1));
        assertArrayEquals(new String[]{"a", "bb", "b"}, APQuiz01.wordsWithoutList(new String[]{"a", "bb", "b", "ccc"}, 3));
        assertArrayEquals(new String[]{"a", "bb", "b", "ccc"}, APQuiz01.wordsWithoutList(new String[]{"a", "bb", "b", "ccc"}, 4));
        assertArrayEquals(new String[]{"xx", "yyy", "yy"}, APQuiz01.wordsWithoutList(new String[]{"xx", "yyy", "x", "yy", "z"}, 1));
        assertArrayEquals(new String[]{"yyy", "x", "z"}, APQuiz01.wordsWithoutList(new String[]{"xx", "yyy", "x", "yy", "z"}, 2));
    }

    /**
     * < Q8 >
     * Tests the hasOne method.
     */
    @Test
    public void testHasOne() {
        assertTrue(APQuiz01.hasOne(10));
        assertFalse(APQuiz01.hasOne(22));
        assertFalse(APQuiz01.hasOne(220));
        assertTrue(APQuiz01.hasOne(212));
        assertTrue(APQuiz01.hasOne(1));
        assertFalse(APQuiz01.hasOne(9));
        assertTrue(APQuiz01.hasOne(211112));
        assertTrue(APQuiz01.hasOne(121121));
        assertFalse(APQuiz01.hasOne(222222));
        assertTrue(APQuiz01.hasOne(56156));
        assertFalse(APQuiz01.hasOne(56556));
    }

    /**
     * < Q9 >
     * Tests the dividesSelf method.
     */
    @Test
    public void testDividesSelf() {
        assertTrue(APQuiz01.dividesSelf(128));
        assertTrue(APQuiz01.dividesSelf(12));
        assertFalse(APQuiz01.dividesSelf(120));
        assertTrue(APQuiz01.dividesSelf(122));
        assertFalse(APQuiz01.dividesSelf(13));
        assertFalse(APQuiz01.dividesSelf(32));
        assertTrue(APQuiz01.dividesSelf(22));
        assertFalse(APQuiz01.dividesSelf(42));
        assertTrue(APQuiz01.dividesSelf(212));
        assertFalse(APQuiz01.dividesSelf(213));
        assertTrue(APQuiz01.dividesSelf(162));
    }

    /**
     * < Q10 >
     * Tests the copyEvens method.
     */
    @Test
    public void testCopyEvens() {
        assertArrayEquals(new int[]{2, 4}, APQuiz01.copyEvens(new int[]{3, 2, 4, 5, 8}, 2));
        assertArrayEquals(new int[]{2, 4, 8}, APQuiz01.copyEvens(new int[]{3, 2, 4, 5, 8}, 3));
        assertArrayEquals(new int[]{6, 2, 4}, APQuiz01.copyEvens(new int[]{6, 1, 2, 4, 5, 8}, 3));
        assertArrayEquals(new int[]{6, 2, 4, 8}, APQuiz01.copyEvens(new int[]{6, 1, 2, 4, 5, 8}, 4));
        assertArrayEquals(new int[]{4}, APQuiz01.copyEvens(new int[]{3, 1, 4, 1, 5}, 1));
        assertArrayEquals(new int[]{2}, APQuiz01.copyEvens(new int[]{2}, 1));
        assertArrayEquals(new int[]{6, 2}, APQuiz01.copyEvens(new int[]{6, 2, 4, 8}, 2));
        assertArrayEquals(new int[]{6, 2, 4}, APQuiz01.copyEvens(new int[]{6, 2, 4, 8}, 3));
        assertArrayEquals(new int[]{6, 2, 4, 8}, APQuiz01.copyEvens(new int[]{6, 2, 4, 8}, 4));
        assertArrayEquals(new int[]{8}, APQuiz01.copyEvens(new int[]{1, 8, 4}, 1));
        assertArrayEquals(new int[]{8, 4}, APQuiz01.copyEvens(new int[]{1, 8, 4}, 2));
        assertArrayEquals(new int[]{2, 8}, APQuiz01.copyEvens(new int[]{2, 8, 4}, 2));
    }

    /**
     * < Q11 >
     * Tests the factorial method.
     */
    @Test
    public void testFactorial() {
        assertEquals(1, APQuiz01.factorial(1));
        assertEquals(2, APQuiz01.factorial(2));
        assertEquals(6, APQuiz01.factorial(3));
        assertEquals(24, APQuiz01.factorial(4));
        assertEquals(120, APQuiz01.factorial(5));
        assertEquals(720, APQuiz01.factorial(6));
        assertEquals(5040, APQuiz01.factorial(7));
        assertEquals(40320, APQuiz01.factorial(8));
        assertEquals(479001600, APQuiz01.factorial(12));
    }

    /**
     * < Q12 >
     * Tests the bunnyEars method.
     */
    @Test
    public void testBunnyEars() {
        assertEquals(0, APQuiz01.bunnyEars(0));
        assertEquals(2, APQuiz01.bunnyEars(1));
        assertEquals(4, APQuiz01.bunnyEars(2));
        assertEquals(6, APQuiz01.bunnyEars(3));
        assertEquals(8, APQuiz01.bunnyEars(4));
        assertEquals(10, APQuiz01.bunnyEars(5));
        assertEquals(24, APQuiz01.bunnyEars(12));
        assertEquals(100, APQuiz01.bunnyEars(50));
        assertEquals(468, APQuiz01.bunnyEars(234));
    }

    /**
     * < Q13 >
     * Tests the fibonacci method.
     */
    @Test
    public void testFibonacci() {
        assertEquals(0, APQuiz01.fibonacci(0));
        assertEquals(1, APQuiz01.fibonacci(1));
        assertEquals(1, APQuiz01.fibonacci(2));
        assertEquals(2, APQuiz01.fibonacci(3));
        assertEquals(3, APQuiz01.fibonacci(4));
        assertEquals(5, APQuiz01.fibonacci(5));
        assertEquals(8, APQuiz01.fibonacci(6));
        assertEquals(13, APQuiz01.fibonacci(7));
    }

    /**
     * < Q14 >
     * Tests the bunnyEars2 method.
     */
    @Test
    public void testBunnyEars2() {
        assertEquals(0, APQuiz01.bunnyEars2(0));
        assertEquals(2, APQuiz01.bunnyEars2(1));
        assertEquals(5, APQuiz01.bunnyEars2(2));
        assertEquals(7, APQuiz01.bunnyEars2(3));
        assertEquals(10, APQuiz01.bunnyEars2(4));
        assertEquals(12, APQuiz01.bunnyEars2(5));
        assertEquals(15, APQuiz01.bunnyEars2(6));
        assertEquals(25, APQuiz01.bunnyEars2(10));
    }
}

File: prev/ap-quiz-01.test.js
/**
 * ap-quiz-01.test.js
 *
 * JUnit-style tests for the functions in ap-quiz-01.js
 *
 * INSTRUCTIONS:
 *     This file provides Jest tests for the methods in the ap-quiz-01.js module.
 *
 * DEPENDENCIES:
 *     This test file requires Jest testing framework.
 *     Install dependencies:
 *         npm init -y
 *         npm install --save-dev jest
 *
 * COMPILE & EXECUTE:
 *
 *     Run all tests:
 *         npm test
 *         OR
 *         npx jest ap-quiz-01.test.js
 *
 * TIP:
 *     - Make sure the require path matches your project structure
 *     - If the file is in the same directory, use: require('./ap-quiz-01')
 *     - Jest automatically finds files ending in .test.js or .spec.js
 */
const {
  scoresIncreasing,
  scores100,
  scoresClump,
  scoresAverage,
  wordsCount,
  wordsFront,
  wordsWithoutList,
  hasOne,
  dividesSelf,
  copyEvens,
  factorial,
  bunnyEars,
  fibonacci,
  bunnyEars2
} = require('./ap-quiz-01');

describe('APQuiz01', () => {

  // Q1
  test('scoresIncreasing', () => {
    expect(scoresIncreasing([1, 3, 4])).toBe(true);
    expect(scoresIncreasing([1, 3, 2])).toBe(false);
    expect(scoresIncreasing([1, 1, 4])).toBe(true);
    expect(scoresIncreasing([1, 1, 2, 4, 4, 7])).toBe(true);
    expect(scoresIncreasing([1, 1, 2, 4, 3, 7])).toBe(false);
    expect(scoresIncreasing([-5, 4, 11])).toBe(true);
  });

  // Q2
  test('scores100', () => {
    expect(scores100([1, 100, 100])).toBe(true);
    expect(scores100([1, 100, 99, 100])).toBe(false);
    expect(scores100([100, 1, 100, 100])).toBe(true);
    expect(scores100([100, 1, 100, 1])).toBe(false);
    expect(scores100([1, 2, 3, 4, 5])).toBe(false);
    expect(scores100([1, 2, 100, 4, 5])).toBe(false);
  });

  // Q3
  test('scoresClump', () => {
    expect(scoresClump([3, 4, 5])).toBe(true);
    expect(scoresClump([3, 4, 6])).toBe(false);
    expect(scoresClump([1, 3, 5, 5])).toBe(true);
    expect(scoresClump([2, 4, 5, 6])).toBe(true);
    expect(scoresClump([2, 4, 5, 7])).toBe(false);
    expect(scoresClump([2, 4, 4, 7])).toBe(true);
    expect(scoresClump([3, 3, 6, 7, 9])).toBe(false);
    expect(scoresClump([3, 3, 7, 7, 9])).toBe(true);
    expect(scoresClump([4, 5, 8])).toBe(false);
  });

  // Q4
  test('scoresAverage', () => {
    expect(scoresAverage([2, 2, 4, 4])).toBe(4);
    expect(scoresAverage([4, 4, 4, 2, 2, 2])).toBe(4);
    expect(scoresAverage([3, 4, 5, 1, 2, 3])).toBe(4);
    expect(scoresAverage([5, 6])).toBe(6);
    expect(scoresAverage([5, 4])).toBe(5);
    expect(scoresAverage([5, 4, 5, 6, 2, 1, 2, 3])).toBe(5);
  });

  // Q5
  test('wordsCount', () => {
    expect(wordsCount(['a', 'bb', 'b', 'ccc'], 1)).toBe(2);
    expect(wordsCount(['a', 'bb', 'b', 'ccc'], 3)).toBe(1);
    expect(wordsCount(['a', 'bb', 'b', 'ccc'], 4)).toBe(0);
    expect(wordsCount(['xx', 'yyy', 'x', 'yy', 'z'], 1)).toBe(2);
    expect(wordsCount(['xx', 'yyy', 'x', 'yy', 'z'], 2)).toBe(2);
    expect(wordsCount(['xx', 'yyy', 'x', 'yy', 'z'], 3)).toBe(1);
  });

  // Q6
  test('wordsFront', () => {
    expect(wordsFront(['a', 'b', 'c', 'd'], 1)).toEqual(['a']);
    expect(wordsFront(['a', 'b', 'c', 'd'], 2)).toEqual(['a', 'b']);
    expect(wordsFront(['a', 'b', 'c', 'd'], 3)).toEqual(['a', 'b', 'c']);
    expect(wordsFront(['a', 'b', 'c', 'd'], 4)).toEqual(['a', 'b', 'c', 'd']);
    expect(wordsFront(['Hi', 'There'], 1)).toEqual(['Hi']);
    expect(wordsFront(['Hi', 'There'], 2)).toEqual(['Hi', 'There']);
  });

  // Q7
  test('wordsWithoutList', () => {
    expect(wordsWithoutList(['a', 'bb', 'b', 'ccc'], 1)).toEqual(['bb', 'ccc']);
    expect(wordsWithoutList(['a', 'bb', 'b', 'ccc'], 3)).toEqual(['a', 'bb', 'b']);
    expect(wordsWithoutList(['a', 'bb', 'b', 'ccc'], 4)).toEqual(['a', 'bb', 'b', 'ccc']);
    expect(wordsWithoutList(['xx', 'yyy', 'x', 'yy', 'z'], 1)).toEqual(['xx', 'yyy', 'yy']);
    expect(wordsWithoutList(['xx', 'yyy', 'x', 'yy', 'z'], 2)).toEqual(['yyy', 'x', 'z']);
  });

  // Q8
  test('hasOne', () => {
    expect(hasOne(10)).toBe(true);
    expect(hasOne(22)).toBe(false);
    expect(hasOne(220)).toBe(false);
    expect(hasOne(212)).toBe(true);
    expect(hasOne(1)).toBe(true);
    expect(hasOne(9)).toBe(false);
    expect(hasOne(211112)).toBe(true);
    expect(hasOne(121121)).toBe(true);
    expect(hasOne(222222)).toBe(false);
    expect(hasOne(56156)).toBe(true);
    expect(hasOne(56556)).toBe(false);
  });

  // Q9
  test('dividesSelf', () => {
    expect(dividesSelf(128)).toBe(true);
    expect(dividesSelf(12)).toBe(true);
    expect(dividesSelf(120)).toBe(false);
    expect(dividesSelf(122)).toBe(true);
    expect(dividesSelf(13)).toBe(false);
    expect(dividesSelf(32)).toBe(false);
    expect(dividesSelf(22)).toBe(true);
    expect(dividesSelf(42)).toBe(false);
    expect(dividesSelf(212)).toBe(true);
    expect(dividesSelf(213)).toBe(false);
    expect(dividesSelf(162)).toBe(true);
  });

  // Q10
  test('copyEvens', () => {
    expect(copyEvens([3, 2, 4, 5, 8], 2)).toEqual([2, 4]);
    expect(copyEvens([3, 2, 4, 5, 8], 3)).toEqual([2, 4, 8]);
    expect(copyEvens([6, 1, 2, 4, 5, 8], 3)).toEqual([6, 2, 4]);
    expect(copyEvens([6, 1, 2, 4, 5, 8], 4)).toEqual([6, 2, 4, 8]);
    expect(copyEvens([3, 1, 4, 1, 5], 1)).toEqual([4]);
    expect(copyEvens([2], 1)).toEqual([2]);
    expect(copyEvens([6, 2, 4, 8], 2)).toEqual([6, 2]);
    expect(copyEvens([6, 2, 4, 8], 3)).toEqual([6, 2, 4]);
    expect(copyEvens([6, 2, 4, 8], 4)).toEqual([6, 2, 4, 8]);
    expect(copyEvens([1, 8, 4], 1)).toEqual([8]);
    expect(copyEvens([1, 8, 4], 2)).toEqual([8, 4]);
    expect(copyEvens([2, 8, 4], 2)).toEqual([2, 8]);
  });

  // Q11
  test('factorial', () => {
    expect(factorial(1)).toBe(1);
    expect(factorial(2)).toBe(2);
    expect(factorial(3)).toBe(6);
    expect(factorial(4)).toBe(24);
    expect(factorial(5)).toBe(120);
    expect(factorial(6)).toBe(720);
    expect(factorial(7)).toBe(5040);
    expect(factorial(8)).toBe(40320);
    expect(factorial(12)).toBe(479001600);
  });

  // Q12
  test('bunnyEars', () => {
    expect(bunnyEars(0)).toBe(0);
    expect(bunnyEars(1)).toBe(2);
    expect(bunnyEars(2)).toBe(4);
    expect(bunnyEars(3)).toBe(6);
    expect(bunnyEars(4)).toBe(8);
    expect(bunnyEars(5)).toBe(10);
    expect(bunnyEars(12)).toBe(24);
    expect(bunnyEars(50)).toBe(100);
    expect(bunnyEars(234)).toBe(468);
  });

  // Q13
  test('fibonacci', () => {
    expect(fibonacci(0)).toBe(0);
    expect(fibonacci(1)).toBe(1);
    expect(fibonacci(2)).toBe(1);
    expect(fibonacci(3)).toBe(2);
    expect(fibonacci(4)).toBe(3);
    expect(fibonacci(5)).toBe(5);
    expect(fibonacci(6)).toBe(8);
    expect(fibonacci(7)).toBe(13);
  });

  // Q14
  test('bunnyEars2', () => {
    expect(bunnyEars2(0)).toBe(0);
    expect(bunnyEars2(1)).toBe(2);
    expect(bunnyEars2(2)).toBe(5);
    expect(bunnyEars2(3)).toBe(7);
    expect(bunnyEars2(4)).toBe(10);
    expect(bunnyEars2(5)).toBe(12);
    expect(bunnyEars2(6)).toBe(15);
    expect(bunnyEars2(10)).toBe(25);
  });

});
